### PR TITLE
(Towards #1618) add transformation to convert InvokeCall to Call.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
This PR essentially takes the code that 'lowered' an InvokeCall into a Call out of the `lower_to_language_level` method and puts it in a new transformation. This then makes it possible to generate true Algorithm layer code (complete with `call invoke(...)`) or to create transformed algorithm code containing standard calls to PSy-layer subroutines. As part of this I've also renamed the existing 'raising' transformation to hopefully make things a bit clearer.

If the CI is green then this is ready to go. Probably one for @rupertford or perhaps @sergisiso.

**Review by rupertford — CHANGES_REQUESTED** [review]
Looks great. The tests are failing so could you please take a look. I've made some comments inline but they are pretty much cosmetic.

File: src/psyclone/domain/common/transformations/alg_invoke_2_psy_call_trans.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: R. W. Ford and A. R. Porter, STFC Daresbury Lab.
+
+''' Transform a PSyclone algorithm-layer-specific invoke call into a call
+to the corresponding PSy-layer routine.
+
+'''
+
+from psyclone.core import SymbolicMaths
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/domain/common/transformations/alg_invoke_2_psy_call_trans.py:41]: Done although we don't have a rule for that says the imported names ought to be in alpha order.

File: src/psyclone/domain/common/transformations/alg_invoke_2_psy_call_trans.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+    '''
+    def validate(self, node, options=None):
+        '''Validate the node argument.
+
+        :param node: a PSyIR node capturing an invoke call.
+        :type node: \
+            :py:class:`psyclone.domain.common.algorithm.AlgorithmInvokeCall`
+        :param options: a dictionary with options for transformations.
+        :type options: dict of str:values or None
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/domain/common/transformations/alg_invoke_2_psy_call_trans.py:64]: I think so. Done.

File: src/psyclone/domain/common/transformations/alg_invoke_2_psy_call_trans.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        ''' Apply the transformation to the supplied AlgorithmInvokeCall.
+        The supplied node will be replaced with a Call node with appropriate
+        arguments. If there are no more invoke calls in the scope of the symbol
+        table containing the 'invoke' symbol then that symbol is removed.
+
+        :param node: a PSyIR algorithm invoke call node.
+        :type node: \
+            :py:class:`psyclone.domain.common.psyir.AlgorithmInvokeCall`
+        :param options: a dictionary with options for transformations.
+        :type options: dict of str:values or None
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/domain/common/transformations/alg_invoke_2_psy_call_trans.py:95]: Done.

File: src/psyclone/tests/domain/common/transformations/alg_invoke_2_psy_call_trans_test.py [hunk 4]
```diff
@@ -0,0 +1,428 @@
...
+            indices = args[idx].children
+            assert len(indices) == len(indices_info)
+            for idx2, index_info in enumerate(indices_info):
+                if isinstance(index_info, str):
+                    assert indices[idx2].symbol.name == index_info
+                else:  # BinaryOperation
+                    assert isinstance(indices[idx2], BinaryOperation)
+
+
+def test_ai2psycall_validate():
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/domain/common/transformations/alg_invoke_2_psy_call_trans_test.py:89]: 'tis done.

File: src/psyclone/domain/common/transformations/alg_trans.py [hunk 5]
```diff
@@ -31,15 +31,15 @@
...
+# Authors: R. W. Ford and A. R. Porter, STFC Daresbury Lab
 
 '''Specialise generic PSyIR representing an algorithm layer to a
 PSyclone algorithm-layer-specific PSyIR which uses specialised classes.
 
 '''
 from psyclone.psyGen import Transformation
 from psyclone.psyir.nodes import Call, Routine, Container
-from psyclone.domain.common.transformations import InvokeCallTrans
+from psyclone.domain.common.transformations import RaiseCall2InvokeTrans
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/domain/common/transformations/alg_trans.py:42]: Oops, sorry.

File: src/psyclone/domain/lfric/transformations/lfric_alg_invoke_2_psy_call_trans.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+    '''
+    def validate(self, node, options=None):
+        '''Validate the node argument.
+
+        :param node: a PSyIR node capturing an LFRicinvoke call.
+        :type node: \
+        :py:class:`psyclone.domain.lfric.algorithm.LFRicAlgorithmInvokeCall`
+        :param options: a dictionary with options for transformations.
+        :type options: dict of str:values or None
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/domain/lfric/transformations/lfric_alg_invoke_2_psy_call_trans.py:60]: In fact, I've updated the Dev Guide appropriately too.

File: src/psyclone/tests/domain/lfric/transformations/lfric_alg_invoke_2_psy_call_trans_test.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author A. R. Porter, STFC Daresbury Lab
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/domain/lfric/transformations/lfric_alg_invoke_2_psy_call_trans_test.py:34]: Probably. I've added you.

**Comment by arporter** [comment]
Ready for another look now.

File: src/psyclone/tests/domain/lfric/algorithm/lfric_alg_psyir_test.py [hunk 8]
```diff
@@ -134,49 +134,6 @@ def test_lfricalgorithminvokecall_create_noname():
...
-    assert call1.routine.name == "invoke_test_1"
-    assert call1.routine.is_import
-    assert call1.routine.interface.container_symbol.name == "alg1_psy"
-    args = call1.children
-    assert len(args) == 1
-    assert isinstance(args[0], Reference)
-    assert args[0].symbol.name == "field1"
-
-
 def test_aic_defcontainerrootname():
```
Comment (rupertford) [inline src/psyclone/tests/domain/lfric/algorithm/lfric_alg_psyir_test.py:137]: domain/lfric/algorithm/lfric_alg_psyir_test.py:52:0: W0611: Unused Reference imported from psyclone.psyir.nodes (unused-import)

**Review by rupertford — CHANGES_REQUESTED** [review]
Very nearly there. Just one issue with conformance and a request to modify the documentation. Everything is pycodestyle and pylint clean. docs build. file-to-file coverage is 100%

File: doc/developer_guide/transformations.rst [hunk 9]
```diff
@@ -101,7 +101,35 @@ Raising Transformations for the LFRic API
...
+There are two code-generation options for Algorithm PSyIR:
+
+ 1. generate the transformed version of the algorithm with normal subroutine
+    calls to PSy-layer routines;
+ 2. generate an algorithm layer complete with 'invoke' calls;
+
+In normal PSyclone operation it is the first option that is required. However,
+if PSyclone is being used to generate an Algorithm layer (e.g. when
+constructing a kernel test harness) then the second option is required. In
+order to support both use cases, PSyclone provides a transformation that
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/developer_guide/transformations.rst:118]: Done.

File: src/psyclone/tests/domain/lfric/algorithm/lfric_alg_psyir_test.py (hunk 8 above)
Comment (arporter) [reply src/psyclone/tests/domain/lfric/algorithm/lfric_alg_psyir_test.py:136]: Oops, done.

**Comment by arporter** [comment]
Requested changes made and brought up-to-date with master. Ready for another look assuming CI happiness.

**Review by rupertford — APPROVED** [review]
All issues have been addressed. Approving.

**Pull request merged**